modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/CacheQueryMemoryLeakTest.java

```diff
@@ -0,0 +1,125 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query;
+
+import java.util.Collections;
+import java.util.List;
+import org.apache.ignite.Ignite;
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.cache.QueryEntity;
+import org.apache.ignite.cache.query.Query;
+import org.apache.ignite.cache.query.QueryCursor;
+import org.apache.ignite.cache.query.SqlFieldsQuery;
+import org.apache.ignite.cache.query.annotations.QuerySqlField;
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.IgniteKernal;
+import org.apache.ignite.spi.discovery.tcp.TcpDiscoverySpi;
+import org.apache.ignite.spi.discovery.tcp.ipfinder.vm.TcpDiscoveryVmIpFinder;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+
+/** */
+public class CacheQueryMemoryLeakTest extends GridCommonAbstractTest {
+    /** */
+    private static final TcpDiscoveryVmIpFinder IP_FINDER = new TcpDiscoveryVmIpFinder(true);
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration igniteCfg = super.getConfiguration(igniteInstanceName);
+        ((TcpDiscoverySpi)igniteCfg.getDiscoverySpi()).setIpFinder(IP_FINDER);
+
+        if (igniteInstanceName.equals("client"))
+            igniteCfg.setClientMode(true);
+
+        return igniteCfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+    }
+
+    /**
+     * Check, that query results are not accumulated, when result set size is a multiple of a {@link Query#pageSize}.
+     *
+     * @throws Exception If failed.
+     */
+    public void testResultIsMultipleOfPage() throws Exception {
+        IgniteKernal srv = (IgniteKernal)startGrid("server");
+        Ignite client = startGrid("client");
+
+        IgniteCache<Integer, Person> cache = startPeopleCache(client);
+
+        int pages = 10;
+        int pageSize = 1024;
+
+        for (int i = 0; i < pages * pageSize; i++) {
+            Person p = new Person("Person #" + i, 25);
+            cache.put(i, p);
+        }
+
+        for (int i = 0; i < 10_000; i++) {
+            if (i % 1000 == 0)
+                System.out.println("Select iteration #" + i);
+
+            Query<List<?>> qry = new SqlFieldsQuery("select * from people");
+            qry.setPageSize(pageSize);
+            QueryCursor<List<?>> cursor = cache.query(qry);
+            cursor.getAll();
+            cursor.close();
+
+            assertTrue("Long JVM pause detected", srv.getLongJVMPausesTotalDuration() < 5000);
```

Comment:
Here I think that peeking into some internal memory structure will be better than relying on GC, since we do not know how much RAM is available on e.g. TC agents.